framework/test/mocha/integration/delegates/forge.js

```diff
@@ -0,0 +1,113 @@
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+'use strict';
+
+const util = require('util');
+const { transfer } = require('@liskhq/lisk-transactions');
+const application = require('../../common/application');
+const Bignum = require('../../../../src/modules/chain/helpers/bignum');
+const random = require('../../common/utils/random');
+const accountFixtures = require('../../fixtures/accounts');
+
+const genesisBlock = __testContext.config.genesisBlock;
+const { NORMALIZER } = global.__testContext.config;
+
+const localCommon = require('../common');
+
+const addTransactionsAndForge = util.promisify(
+	localCommon.addTransactionsAndForge
+);
+const transactionInPool = localCommon.transactionInPool;
+const addTransaction = util.promisify(localCommon.addTransaction);
+const fillPool = util.promisify(localCommon.fillPool);
+const forge = util.promisify(localCommon.forge);
+
+let library;
+
+describe('delegates (forge)', () => {
+	before(done => {
+		application.init(
+			{
+				sandbox: {
+					name: 'lisk_test_delegates_forge',
+				},
+			},
+			(err, scope) => {
+				library = scope;
+				library.modules.blocks.lastBlock.set(genesisBlock);
+				done();
+			}
+		);
+	});
+
+	after(done => {
+		application.cleanup(done);
+	});
+
+	describe('forge', () => {
+		describe('total spending', () => {
+			it('should not include transactions which exceed total spending per account balance', async () => {
+				// Transfer 1LSK to account
+				const account = random.account();
+				const transaction = transfer({
+					amount: new Bignum(NORMALIZER).multipliedBy(1).toString(),
+					recipientId: account.address,
+					passphrase: accountFixtures.genesis.passphrase,
+				});
+				await addTransactionsAndForge(library, [transaction], 0);
+
+				// Transfer 0.5 LSK to a random account - The account balance would be 0.4
+				const trs_06 = transfer({
+					amount: new Bignum(NORMALIZER).multipliedBy(0.6).toString(),
+					recipientId: random.account().address,
+					passphrase: account.passphrase,
+				});
+
+				// Transfer 0.5 LSK to a random account - The account balance would be -0.2
+				const trs_04 = transfer({
+					amount: new Bignum(NORMALIZER).multipliedBy(0.4).toString(),
+					recipientId: random.account().address,
+					passphrase: account.passphrase,
+				});
+
+				// Transfer 0.3 LSK to that account - The account balance would be 0.1
+				const trs_03 = transfer({
+					amount: new Bignum(NORMALIZER).multipliedBy(0.3).toString(),
+					recipientId: account.address,
+					passphrase: accountFixtures.genesis.passphrase,
+				});
+
+				// Add the transactions to pool and forge
+				await addTransaction(library, trs_06);
+				await addTransaction(library, trs_04);
+				await addTransaction(library, trs_03);
+				await fillPool(library);
```

Comment:
This line shouldn't be here, as `fillPool` is called inside of `forge`. However, when we remove this call and we have in the transaction pool a transaction that should not be included in a block because of `_verifyTotalSpending` - the empty block is forged. The block should not be empty, it should just not contain the invalid transaction. This is a regression because in earlier versions the behavior was correct, I'm not sure if this is a result of this pull request or other changes.

Reply:
@4miners I tested it yesterday, there was difference of behaviour in `forge` add `addTransactionAndForge`, the later was not working in any case, not even with single transactions. I didn't debugged it why its not, but used `fillPool` and `forge` separately in this test.